src/docfx/template/TemplateEngine.cs

```diff
@@ -12,47 +13,79 @@ namespace Microsoft.Docs.Build
 {
     internal class TemplateEngine : IDisposable
     {
-        private readonly string _templateDir;
-        private readonly string _schemaDir;
-        private readonly string _contentTemplateDir;
-        private readonly JObject _global;
-        private readonly LiquidTemplate _liquid;
-        private readonly ThreadLocal<IJavaScriptEngine> _js;
-        private readonly MustacheTemplate _mustacheTemplate;
+        private readonly string? _templateDir;
+        private readonly string? _schemaDir;
+        private readonly string? _contentTemplateDir;
+        private readonly JObject? _global;
+        private readonly LiquidTemplate? _liquid;
+        private readonly ThreadLocal<IJavaScriptEngine>? _js;
+        private readonly MustacheTemplate? _mustacheTemplate;
+        private readonly TemplateDefinition? _templateDefinition;
+        private readonly FileResolver _fileResolver;
 
         private readonly ConcurrentDictionary<string, JsonSchemaValidator?> _schemas
                    = new ConcurrentDictionary<string, JsonSchemaValidator?>(StringComparer.OrdinalIgnoreCase);
 
-        public TemplateEngine(Config config, BuildOptions buildOptions, PackageResolver packageResolver, Lazy<JsonSchemaTransformer> jsonSchemaTransformer)
```

Comment:
We probably want to extract a TemplateProvider and have two implementations for package-based and file-based.